### ISSUE TITLE
Add a howto for "Deploying Twisted with SystemD"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5601]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5601|
|Type|enhancement|
|Created|2012-04-02 20:19:09Z|
|Branch|https://github.com/twisted/twisted/tree/systemd-howto-5601|

[#5575](https://github.com/twisted/twisted/issues/5575) adds a systemd endpoint for setting up servers using sockets inherited from systemd.  This requires some configuration, though.  We should have a brief howto pointing out the basics necessary for using this feature.

Attachments:

* [5601-systemd-documentation.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/47c/47c0d93b1ba943376ca72df29ad4674ab3e15201/c0bfd7557fa2d56901ce9dfc624e89047f557fca.patch) (4984 bytes) - added by rwall on 2012-04-07 01:42:14Z - an example of systemd socket activation of twistd web
* [systemd-howto-5601-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/47c/47c0d93b1ba943376ca72df29ad4674ab3e15201/8e6bc2a557873b7a9cce81a2b08beb54ff5f5ffb.patch) (6987 bytes) - added by rwall on 2013-01-22 23:12:07Z - Updated systemd socket endpoint documentation addressing code review comments from thijs
* [systemd-howto-5601-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/47c/47c0d93b1ba943376ca72df29ad4674ab3e15201/14f545f2de9ede6935e3eeea5d10c4ad402f5921.patch) (22803 bytes) - added by rwall on 2013-01-29 18:28:00Z - Updated patch
* [systemd-howto-5601-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/47c/47c0d93b1ba943376ca72df29ad4674ab3e15201/4b9b2e44b4231677a7518049c04c3e71422fd68c.patch) (16381 bytes) - added by rwall on 2013-02-12 12:09:55Z - Updated patch against source:branches/systemd-howto-5601 addressing code review comments in #comment:8

<details><summary>Searchable metadata</summary>

```
trac-id__5601 5601
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_systemd_howto_5601 branches/systemd-howto-5601
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1333397949000000 1333397949000000
changetime__1360727086000000 1360727086000000
version__None None
owner__tom_prince tom.prince
cc__thijs cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

Thanks for the patch. Putting it up for review, as outlined on ReviewProcess#Authors:Howtogetyourchangereviewed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra set owner to @wallrj|
|-|-|

Thanks for your patch.

 1. this needs a .doc news file as described on ReviewProcess#Newsfiles
 1. Combinator in doc/core/howto/endpoints.xhtml should be linked to [the wikli page](https://twistedmatrix.com/trac/wiki/Combinator).
 1. systemd can use a link to the API documentation
 1. period behind `until the first client connects`
 1. `curl` can use an external link
 1. `--nodaemon`, `PYTHONPATH`, and `index=0` should be wrapped in `<code></code>`
 1. twisted should be Twisted

Please diff against the branch for future patches.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

The documentation builder also [failed](http://buildbot.twistedmatrix.com/builders/documentation/builds/2040/steps/process-docs/logs/stdio).

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj removed owner|
|-|-|

Ready for another review in [systemd-howto-5601-2.patch](None/systemd-howto-5601-2.patch)

 * I addressed all of thijs comments
 * fixed the broken xhtml
 * added some further links
 * re-tested on Fedora 17

 * Note: The patch does contain a topfile but its empty and doesn't show up in the trac diff viewer.
 * Note: This patch is against trunk not the branch above.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @wallrj<br>@exarkun set status to `new`|
|-|-|

Thanks for your continued work on this.  I'm pretty happy we might get docs for using systemd support in less than a year after getting systemd support in. ;)

  1. I think this whole section added to endpoints.xhtml would be better as the start of a brand new document about deploying Twisted applications.  The brief systemd example already in endpoints.xhtml can link people to the new document.  Adding all of this systemd material to the endpoints document is going to mostly be presenting it to the wrong audience.
  1. I'm not sure it's a good idea to involve Combinator in the example.  Hardly anyone has any idea what Combinator is.  Practically no one is going to want to deploy a Twisted service using it, let alone a systemd managed service (I suspect).  If the idea is to demonstrate setting something in the environment, then something more generally known or used might be better - perhaps `OPENSSL_CONF` (which I think Twisted's SSL support will even respect (by accident)).  Something even more widely known and simple would be better, but I can't think of any other interesting environment variables from the real world right now.
  1. This is a pretty big chunk of nice documentation.  I think it merits a topfile with some contents, rather than just a misc!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj removed owner|
|-|-|

Ready for another review in  [systemd-howto-5601-3.patch](None/systemd-howto-5601-3.patch)
 * Moved all the systemd documentation to a separate file
 * Linked to it from the core deployment overview page
 * Separate sections for starting a twisted service statically
 * ...and starting a twisted service with socket activation.

This patch is against trunk.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince set owner to @wallrj|
|-|-|

Thanks for your work on this.

  1. Man page urls should probably point at [freedesktop.org](http://www.freedesktop.org/software/systemd/man/) rather than at [0pointer.de](http://0pointer.de/public/systemd-man/).
  1. `sudo` probably isn't needed for status.
  1. The `[Install]` section is implicitly reference (in the discussion of the symlink created by `systemd enable`), but nowhere explained.
  1. It might be worth pointing out that using socket activation allows you bind privileged ports (i.e. <1024) while running as non-root.
  1. There is an [error](https://buildbot.twistedmatrix.com/builders/documentation/builds/2919/steps/process-docs/logs/stdio) building documentation. I think this is probably due to the heading `Create a systemd.service file` containing a link. (Please file a ticket for getting lore to support this for latex)

Please resubmit, with a patch against the branch, after addressing the above point.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj removed owner|
|-|-|

Replying to [tom.prince](#note_8):
> Thanks for your work on this.
>

>   1. Man page urls should probably point at
>   [freedesktop.org](http://www.freedesktop.org/software/systemd/man/)
>   rather than at [http://0pointer.de/public/systemd-man/
>   0pointer.de].

Yep, I agree. Done.

>   1. `sudo` probably isn't needed for status.

Done.

>   1. The `[Install]` section is implicitly reference (in the discussion of the symlink created by `systemd enable`), but nowhere explained.

Done.


>   1. It might be worth pointing out that using socket activation allows you bind privileged ports (i.e. <1024) while running as non-root.

Added a paragraph to that effect and reformatted the socekt activation
quote to make it more readable. It isn't actually a very readable
explanation, but I couldn't think of anything better.

>   1. There is an [error](https://buildbot.twistedmatrix.com/builders/documentation/builds/2919/steps/process-docs/logs/stdio) building documentation. I think this is probably due to the heading `Create a systemd.service file` containing a link. (Please file a ticket for getting lore to support this for latex)

Lore seems to build the docs fine for me...

```
[richard@zorin doc]$ ../bin/lore/lore --config template=core/howto/template.tpl
[########################################] (*Done*)                            s)h)rred)
```

...maybe the buildbot has a different version of lore.

Anyway, I removed the links from the headings - hopefully that will
fix the problem.

> Please resubmit, with a patch against the branch, after addressing the above point.

 * I also deleted a couple of old listings files which were introduced to
   this branch by a previous patch.
 * Used the word "directive" instead of "option"

See [systemd-howto-5601-4.patch](None/systemd-howto-5601-4.patch)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_10"></a>|@wallrj commented|
|-|-|

One other thing. I think the topfile has wrong extension. Should be...

twisted/topfiles/5601.feature => twisted/topfiles/5601.doc

...can the reviewer / merger do that for me in the branch?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to @tomprince|
|-|-|

One last point,
  1. In the see also, link to the main systemd page: [http://www.freedesktop.org/wiki/Software/systemd/] rather than the list of man page.

I'll merge with this change.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set status to `closed`|
|-|-|

(In [37132]) Merge systemd-howto-5601: Add a howto for "Deploying Twisted with SystemD"

Author: rwall
Reviewers: thijs, exarkun, tom.prince
Fixes: [#5601](https://github.com/twisted/twisted/issues/5601)

A new "Deploying Twisted with systemd" howto document which
demonstrates how to start a Twisted service using systemd socket
activation.
